sql/spanner/driver.go

```diff
@@ -0,0 +1,139 @@
+// Copyright 2021-present The Atlas Authors. All rights reserved.
+// This source code is licensed under the Apache 2.0 license found
+// in the LICENSE file in the root directory of this source tree.
+
+package spanner
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"net/url"
+	"os"
+	"path/filepath"
+	"strconv"
+	"strings"
+	"time"
+
+	"ariga.io/atlas/sql/internal/sqlx"
+	"ariga.io/atlas/sql/migrate"
+	"ariga.io/atlas/sql/schema"
+	"ariga.io/atlas/sql/sqlclient"
+	_ "github.com/googleapis/go-sql-spanner"
+)
+
+type (
+	// Driver represents a Spanner driver for introspecting database schemas,
+	// generating diff between schema elements and apply migrations changes.
+	Driver struct {
+		conn
+		schema.Differ
+		schema.Inspector
+		migrate.PlanApplier
+	}
+
+	// database connection and its information.
+	conn struct {
+		schema.ExecQuerier
+		databaseDialect string
+	}
+)
+
+func init() {
+	sqlclient.Register(
+		"spanner",
+		sqlclient.DriverOpener(Open),
+		sqlclient.RegisterCodec(MarshalHCL, EvalHCL),
+		sqlclient.RegisterURLParser(sqlclient.URLParserFunc(func(u *url.URL) *sqlclient.URL {
+			uc := &sqlclient.URL{URL: u, DSN: strings.TrimPrefix(u.String(), u.Scheme+"://")}
+			return uc
+		})),
+	)
+}
+
+// Open opens a new Spanner driver.
+func Open(db schema.ExecQuerier) (migrate.Driver, error) {
+	var (
+		c   = conn{ExecQuerier: db}
+		ctx = context.Background()
+	)
+	ctx, cancel := context.WithTimeout(ctx, 30*time.Second)
+	defer cancel()
+	rows, err := db.QueryContext(ctx, paramsQuery)
+	if err != nil {
+		return nil, fmt.Errorf("spanner: query database options: %w", err)
+	}
+	if err := sqlx.ScanOne(rows, &c.databaseDialect); err != nil {
+		return nil, fmt.Errorf("spanner: query database options: %w", err)
+	}
+	if err := rows.Err(); err != nil {
+		return nil, fmt.Errorf("failed to execute query: %v", err)
+	}
+	return &Driver{
+		conn:      c,
+		Inspector: &inspect{c},
+		// Differ:      &sqlx.Diff{DiffDriver: &diff{c}},
```

Comment:
Delete unused comments

Reply:
I commented this out as we'll be adding/re-adding them shortly -- but that's fair, will trim down.